File: needserver/needserver-core/src/main/java/won/server/service/ConnectionService.java
/*
 * Copyright 2012  Research Studios Austria Forschungsges.m.b.H.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package won.server.service;

import java.net.URI;

/**
 * User: fkleedorfer
 * Date: 23.10.12
 */
public interface ConnectionService
{
  /**
   * OwnerService-facing method; accepts a connection initiated by a connect().
   * @param connectionURI the URI of the connection
   */
  public void accept(URI connectionURI);

  /**
   * OwnerService-facing method; deny a connection initiated by a connect().
   * @param connectionURI the URI of the connection
   */
  public void deny(URI connectionURI);

  /**
   * OwnerService-facing method; Aborts the connection identified by the specified URI, indicating failure.
   * @param connectionURI the URI of the connection
   */
  public void abort(URI connectionURI);

  /**
   * OwnerService-facing method; closes the connection identified by the specified URI, indicating success.
   * @param connectionURI the URI of the connection
   */
  public void close(URI connectionURI);


  /**
   * OwnerService-facing method; sends a chat message via the local connection identified by the specified connectionURI
   * to the remote partner.
   * @param connectionURI the local connection
   * @param message the chat message
   */
  public void sendMessage(URI connectionURI, String message);

  /**
   * NeedService-facing (i.e. connection-facing) method; Informs the connection object of the fact that the connection
   * has been accepted by the other side.
   * @param connectionURI the URI of the connection
   */
  public void connectionAccepted(URI connectionURI);

  /**
   * NeedService-facing (i.e. connection-facing) method; Informs the connection object of the fact that the connection
   * has been denied by the other side.
   * @param connectionURI the URI of the connection
   */
  public void connectionDenied(URI connectionURI);

  /**
   * NeedService-facing (i.e. connection-facing) method; Informs the connection object of the fact that the connection
   * has been aborted by the other side, indicating failure.
   * @param connectionURI the URI of the connection
   */
  public void connectionAborted(URI connectionURI);

  /**
   * NeedService-facing (i.e. connection-facing) method; Informs the connection object of the fact that the connection
   * has been closed by the other side, indicating success.
   * @param connectionURI the URI of the connection
   */
  public void connectionClosed(URI connectionURI);

  /**
   * NeedService-facing method; receives a chat message from the remote connection.
   * to the remote partner.
   * @param connectionURI the local connection
   * @param message the chat message received from the remote connection
   */
  public void messageReceived(URI connectionURI, String message);


}

File: webofneeds/won-owner-webapp/src/main/webapp/app/service/connection-service.js
/*
 * Copyright 2012  Research Studios Austria Forschungsges.m.b.H.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

angular.module('won.owner').factory('connectionService', function ($http) {

	var connectionService = {};

	connectionService.STATE_REQUEST = 'manageConnection'
	connectionService.STATE_CLOSED = 'showMessageClosed'
	connectionService.STATE_CHAT = 'listMessages'
	connectionService.STATE_PENDING = 'showMessagePending'

	connectionService.getConnectionState = function (connectionId) {
		return $http.get('/owner/rest/connection/' + connectionId + '/state').then(function(result) {
			if(result.data.indexOf(connectionService.STATE_CLOSED) > -1) {
				return {
					state :connectionService.STATE_CLOSED,
					message :result.data.split('-')[1]
				}
			} else if(result.data.indexOf(connectionService.STATE_PENDING) > -1) {
				return {
					state:connectionService.STATE_PENDING,
					message:result.data.split('-')[1]
				}
			} else if (result.data.indexOf(connectionService.STATE_REQUEST) > -1) {
				return {
					state:connectionService.STATE_REQUEST
				}
			} else if (result.data.indexOf(connectionService.STATE_CHAT) > -1) {
				return {
					state:connectionService.STATE_CHAT
				}
			} else {
				console.log(result)
			}
		});
	};

	connectionService.getConnectionMessages = function (connectionId) {
		return $http.get('/owner/rest/connection/' + connectionId + '/body');
	};

	connectionService.sendConnectionAccept= function (connectionId) {
		return $http.get('/owner/rest/connection/' + connectionId + '/accept');
	};

	connectionService.sendConnectionDeny = function (connectionId) {
		return $http.get('/owner/rest/connection/' + connectionId + '/deny');
	};

	connectionService.sendConnectionClose = function (connectionId) {
		return $http.get('/owner/rest/connection/' + connectionId + '/close');
	};

	connectionService.sendConnectionMessage = function (connectionId, textMessage) {
		return $http({
			method:'POST',
			url:'http://localhost:8080/owner/rest/connection/' + connectionId + '/send',
			data:textMessage
		})
	};

	return connectionService;
});

